### PR TITLE
BBM: Fix incorrect freecount when there are bad blocks

### PR DESCRIPTION
**Pull request by Jun-Tao**

### CODE REVIEW

File: drivers/mtd/w25.c [hunk 1]
```diff
@@ -1393,13 +1394,14 @@ static int w25_erase(FAR struct mtd_dev_s *dev, off_t startsector, size_t nsecto
 #endif
 }
 
-static int w25_erase2(FAR struct w25_dev_s *priv, off_t startblock, size_t nblocks)
+static int w25_erase2(FAR struct w25_dev_s *priv, off_t startblock, size_t nblocks, uint8_t *pfreecount)
```
Comment (haiyangyuneec) [inline drivers/mtd/w25.c]: pfreecount -> freecount

File: drivers/mtd/w25.c [hunk 2]
```diff
@@ -1415,6 +1417,10 @@ static int w25_erase2(FAR struct w25_dev_s *priv, off_t startblock, size_t nbloc
 
       ret = w25_blockerase(priv, startblock);
       if (ret < 0) {
+        pfreecount[startblock << shift] = SMARTFS_FREECOUNT_BADBLOCK;
+        pfreecount[(startblock << shift) + 1] = SMARTFS_FREECOUNT_BADBLOCK;
+        pfreecount[(startblock << shift) + 2] = SMARTFS_FREECOUNT_BADBLOCK;
+        pfreecount[(startblock << shift) + 3] = SMARTFS_FREECOUNT_BADBLOCK;
```
Comment (haiyangyuneec) [inline drivers/mtd/w25.c]: memset(pfreecount, (1 << shift), SMARTFS_FREECOUNT_BADBLOCK)

In this way, you needn't change code if W25_SECTOR_SHIFT is changed.

File: drivers/mtd/w25.c [hunk 3]
```diff
@@ -276,7 +277,7 @@ static ssize_t w25_byteread(FAR struct w25_dev_s *priv, off_t offset,
 static ssize_t w25_pageread(FAR struct w25_dev_s *priv, off_t address,
                           size_t nbytes, bool spare, FAR uint8_t *buffer);
 #ifndef CONFIG_W25_READONLY
-static int w25_erase2(FAR struct w25_dev_s *priv, off_t startblock, size_t nblocks);
+static int w25_erase2(FAR struct w25_dev_s *priv, off_t startblock, size_t nblocks, uint8_t *pfreecount);
```
Comment (haiyangyuneec) [inline drivers/mtd/w25.c]: pfreecount -> freecount

File: drivers/mtd/w25.c [hunk 4]
```diff
@@ -62,6 +62,7 @@
 //#define CONFIG_W25_CACHE_DEBUG
 //#define CONFIG_W25_API_DEBUG
 
+#define SMARTFS_FREECOUNT_BADBLOCK  0xee
```
Comment (haiyangyuneec) [inline drivers/mtd/w25.c]: Move this into include/nuttx/mtd/mtd.h and remove this from smartfs. Both w25.h and smartfs could refer to it.

File: drivers/mtd/smart.c [hunk 5]
```diff
@@ -1659,32 +1683,12 @@ static inline int smart_llformat(FAR struct smart_struct_s *dev, unsigned long a
...
-#ifdef CONFIG_MTD_SMART_LOGICAL_SECTOR
-      dev->releasecount[x] = prerelease;
-#endif
-      dev->freecount[x] = dev->availSectPerBlk-prerelease;
-    }
 
   /* Account for the format sector */
 
-  if (dev->freecount[dev->rootphyssector] > 0)
+  if ((dev->freecount[dev->rootphyssector] > 0) && (dev->freecount[dev->rootphyssector] != SMART_FREECOUNT_BADBLOCK)) {
```
Comment (haiyangyuneec) [inline drivers/mtd/smart.c]: Wrap to two lines to get shorter line each.

File: drivers/mtd/smart.c [hunk 6]
```diff
@@ -1701,8 +1705,9 @@ static inline int smart_llformat(FAR struct smart_struct_s *dev, unsigned long a
 #ifdef CONFIG_MTD_SMART_LOGICAL_SECTOR
   dev->sMap[SMART_METADATA_SECTOR] = 1;
 #endif
-  if (dev->freecount[SMART_METADATA_SECTOR] > 0)
+  if ((dev->freecount[SMART_METADATA_SECTOR] > 0) && (dev->freecount[SMART_METADATA_SECTOR] != SMART_FREECOUNT_BADBLOCK)) {
```
Comment (haiyangyuneec) [inline drivers/mtd/smart.c]: Wrap to two lines to get shorter line each.

**Review by haiyangyuneec — CHANGES_REQUESTED** [review]

File: include/nuttx/mtd/mtd.h [hunk 7]
```diff
@@ -104,6 +104,8 @@
 #  define CONFIG_MTD_REGISTRATION   1
 #endif
 
+#define SMART_FREECOUNT_BADBLOCK  0xee
```
Comment (haiyangyuneec) [inline include/nuttx/mtd/mtd.h]: I don't think mtd should now about "SMART".  Low level flash driver should not know which file system is using it. How about change it to "MTD_BADBLOCK_MARK"?

Comment (Jun-Tao) [reply include/nuttx/mtd/mtd.h]: OK

**Review by haiyangyuneec — APPROVED** [review]

**Pull request merged**